_source/_posts/2021-09-24-java-records.md

```diff
@@ -0,0 +1,716 @@
+
+```java
+public EndOfGame {
+    Objects.requireNonNull(date);
+    Objects.requireNonNull(timeOfDay);
+    Objects.requireNonNull(mentalState);
+    Objects.requireNonNull(damageTaken);
+    Objects.requireNonNull(damageToPlayers);
+    Objects.requireNonNull(damageToStructures);
+}
+```
+
+In the compact form, the parameters are implicit, the private fields cannot be assigned inside the body and are automatically assigned **at the end**. Its purpose is to allow focusing on validation, making defensive copies of mutable values, or some other value processing.
+
+## Serialization, Encoding, Mapping
+
+Record instances can extend `Serializable`, but the serialization and deserialization processes cannot be customized. Serialization and deserialization methods like `writeObject`,  `readObject` can be implemented, but will be ignored.
+
+As `spring-web` module provides Jackson JSON encoders and decoders, and Java Record support was added to Jackson in release 2.12, records can be used for REST API request and response mapping.
+
+**Records cannot be used as entities with JPA/Hibernate**. JPA entities must have a no-args constructor, and must not be final, two requirements that `record` will not comply, and there are more.
+
+Spring Data modules that do not use the object mapping of the underlying data store (like JPA) do support Record, as persistence construction detection works as with other classes. For Spring Data MongoDB, the general recommendation is to _stick to immutable objects_, because they are straightforward to materialize by calling the constructor. It is also recommended to provide an all-args constructor, allowing to skip property population for optimal performance. Java Record semantics align with these guidelines.
+
+
+# Using Java Record with Spring WebFlux and Spring Data
+
+While searching for a dataset for this tutorial, I came across a collection of 87 end game statistics of a single player, for a popular game. As the author included the mental state in the data, I decided to use Java Records for building a basic average query and finding out if the performance was significantly different when sober than when high.
+
+Let's jump ahead and build the a secured REST Api using the `EndOfGame` record, Spring Boot, [MongoDB](https://www.mongodb.com/community) and Okta authentication. With the help of [Spring Initializr](https://start.spring.io/) create a WebFlux application, from the web UI or with [HTTPie](https://httpie.io/):
+
+```shell
+https -d start.spring.io/starter.zip type==maven-project \
+  language==java \
+  bootVersion==2.5.5 \
+  baseDir==java-records \
+  groupId==com.okta.developer.records \
+  artifactId==records-demo \
+  name==java-records \
+  packageName==com.okta.developer.records \
+  javaVersion==17 \
+  dependencies==webflux,okta,data-mongodb-reactive
```

Comment:
```suggestion
https -d start.spring.io/starter.zip bootVersion==2.5.6 \
  baseDir==java-records \
  groupId==com.okta.developer.records \
  artifactId==records-demo \
  name==java-records \
  packageName==com.okta.developer.records \
  javaVersion==17 \
  dependencies==webflux,okta,data-mongodb-reactive
```
No need to specify the defaults. I upgraded when I QA'd, so will use it in my PR.